test/unit/org/apache/cassandra/metrics/CompactionMetricsTest.java

```diff
@@ -0,0 +1,229 @@
+        {
+            tct.start();
+            final List<CompactionInfo.Holder> activeCompactions = this.getActiveCompactionsForTable(cfs);
+            Assert.assertEquals(1L, activeCompactions.size());
+            Assert.assertEquals(activeCompactions.get(0).getCompactionInfo().getSSTables(), toMarkCompacting);
+            cfs.runWithCompactionsDisabled(() -> null, sstable -> !toMarkCompacting.contains(sstable), false, false, true);
+            Assert.assertEquals(1L, activeCompactions.size());
+            Assert.assertFalse(activeCompactions.get(0).isStopRequested());
+
+            // logger.info("testing data: {} {} {}", initialPendingTasks, CompactionManager.instance.getMetrics().pendingTasks.getValue(), CompactionManager.instance.getMetrics().pendingTasksByTableName.getValue().containsKey(cfs.keyspace.getName()));
+            Assert.assertTrue(CompactionManager.instance.getMetrics().pendingTasksByTableName.getValue().containsKey(cfs.keyspace.getName()));
+            Assert.assertTrue(CompactionManager.instance.getMetrics().pendingTasks.getValue() > initialPendingTasks);
+            logger.info("finish testing");
+        }
+        finally
+        {
+            tct.abort();
+        }
+    }
+
+    @Test
+    public void testCompactionMetricsForFailedTasks()
+    {
+
+        Assert.assertEquals(0, CompactionManager.instance.getMetrics().compactionsAborted.getCount());
+        Assert.assertEquals(0, CompactionManager.instance.getMetrics().compactionsReduced.getCount());
+        Assert.assertEquals(0, CompactionManager.instance.getMetrics().sstablesDropppedFromCompactions.getCount());
+
+
+        final ColumnFamilyStore cfs = MockSchema.newCFS();
+        final List<SSTableReader> sstables = this.createSSTables(cfs, 10, 0);
+        final Set<SSTableReader> toMarkCompacting = new HashSet<>(sstables.subList(0, 3));
+        final CompactionMetricsTest.TestCompactionTask tct = new CompactionMetricsTest.TestCompactionTask(cfs, toMarkCompacting);
+        try
+        {
+            tct.start2();
+            final List<CompactionInfo.Holder> activeCompactions = this.getActiveCompactionsForTable(cfs);
+            Assert.assertEquals(1L, activeCompactions.size());
+            Assert.assertEquals(activeCompactions.get(0).getCompactionInfo().getSSTables(), toMarkCompacting);
+            cfs.runWithCompactionsDisabled(() -> null, sstable -> !toMarkCompacting.contains(sstable), false, false, true);
+            Assert.assertEquals(1L, activeCompactions.size());
+            Assert.assertFalse(activeCompactions.get(0).isStopRequested());
+        }
+        finally
+        {
+            tct.abort();
+        }
+
+        Assert.assertTrue(CompactionManager.instance.getMetrics().compactionsAborted.getCount() > 0);
+        Assert.assertTrue(CompactionManager.instance.getMetrics().compactionsReduced.getCount() > 0);
+        Assert.assertTrue(CompactionManager.instance.getMetrics().sstablesDropppedFromCompactions.getCount() > 0);
```

Comment:
Concern here is that we're just asserting counters that we ticked ourselves below. Ideally we would be exercising the actual production code that hits these. Did you explore mocking or other techniques that might allow us to hit this code?
https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/db/compaction/CompactionTask.java#L394-L409